modules/apps/depot/depot-service/src/main/java/com/liferay/depot/service/impl/DepotEntryPinServiceImpl.java

```diff
@@ -0,0 +1,144 @@
+import com.liferay.depot.model.DepotEntryPin;
+import com.liferay.depot.service.base.DepotEntryPinServiceBaseImpl;
+import com.liferay.portal.aop.AopService;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.security.permission.ActionKeys;
+import com.liferay.portal.kernel.security.permission.resource.ModelResourcePermission;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+/**
+ * @author Brian Wing Shun Chan
+ */
+@Component(
+	property = {
+		"json.web.service.context.name=depot",
+		"json.web.service.context.path=DepotEntryPin"
+	},
+	service = AopService.class
+)
+public class DepotEntryPinServiceImpl extends DepotEntryPinServiceBaseImpl {
+
+	@Override
+	public DepotEntryPin addDepotEntryPin(long userId, long depotEntryId)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.UPDATE);
+
+		return depotEntryPinLocalService.addDepotEntryPin(userId, depotEntryId);
+	}
+
+	@Override
+	public DepotEntryPin deleteDepotEntryPin(long userId, long depotEntryId)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.UPDATE);
+
+		return depotEntryPinLocalService.deleteDepotEntryPin(
+			userId, depotEntryId);
+	}
+
+	@Override
+	public List<DepotEntryPin> getDepotEntryDepotEntryPins(
+			long depotEntryId, int start, int end)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.VIEW);
+
+		return depotEntryPinLocalService.getDepotEntryDepotEntryPins(
+			depotEntryId, start, end);
+	}
+
+	@Override
+	public int getDepotEntryDepotEntryPinsCound(
+			long depotEntryId, int start, int end)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.VIEW);
+
+		return depotEntryPinLocalService.getDepotEntryDepotEntryPinsCount(
+			depotEntryId);
+	}
+
+	@Override
+	public List<DepotEntryPin> getUserDepotEntryPins(
+			long userId, int start, int end)
+		throws PortalException {
+
+		List<DepotEntryPin> depotEntryPins = new ArrayList<>();
+
+		while (depotEntryPins.size() < end) {
+			List<DepotEntryPin> userDepotEntryPins =
+				depotEntryPinPersistence.findByUserId(userId, start, end);
+
+			if (userDepotEntryPins.isEmpty()) {
+				break;
+			}
+
+			for (DepotEntryPin userDepotEntryPin : userDepotEntryPins) {
+				if (_depotEntryModelResourcePermission.contains(
+						getPermissionChecker(),
+						userDepotEntryPin.getDepotEntryId(), ActionKeys.VIEW)) {
+
+					depotEntryPins.add(userDepotEntryPin);
+				}
+
+				start = start + end;
+
+				end = start + end;
```

Comment:
Is this correct? This will increase the page size on each iteration, and it will skip items, leaving holes in the pagination.

I think what you intended was something more like this:
```
int pageSize = end - start; // compute page size at the beginning.

while (...) {
   ...
   start += pageSize;
   end += pageSize;
}
```
Increasing the page size to avoid doing too many queries is fine (even desirable), but leaving holes between pages is wrong.

Reply:
In any case, why do we need to filter the pins? If the user has a pin, it mean he could access it; and if the permission is removed, then the pin should not work anyway (we should be checking permissions when accessing a space).